### ISSUE TITLE
Proposal for improved underline highlighting

### ISSUE THREAD

**Issue by agrogers** [AndBible/and-bible#1407]
The following method resolves the issue of only being able to show a single color underline. It also (elegantly) uses the existing the background highlighting methodology.

It resolves a few issues:

1. Can only show a single color
2. Weird underline style for verse numbers and footnotes
3. Works with the underline used for Strongs numbers
4. Would leave the underline decorator open for other uses

Below is screen shot and CSS of how it should work.

Here is what worked nicely for me:
1. Add a bottom-padding of 0.5em to allow room for the underlines to appear below the text.
2. Allocate 65% of the background gradient to the background highlight colors
3. Allocate 5% to each underline. 
4. Allocate 5% (or less) to each space between the line.

A few things that might be nice.
1. When an underline moves from one verse to another or perhaps between words it would be nice if the line continued on rather than jumping up or down. I think this would be relatively easy to manage - just need to keep track of the previous spans highlighting and ensure that if an underline appears in the current span it is allocated the same index as before.
2. Any number of underlines are possible but i suspect 4 or 5 is about the practical limit... unless the text is made larger and then you can get a lot more in.
3. There was some dark magic going on in the code to calculate these splits. So as much as i would love to try and implement this I don't think I have the sufficient powers :)

![image](https://user-images.githubusercontent.com/13920678/126884557-c652b3fc-bc68-47cd-b4fc-8a393bbc0dda.png)

```
    padding-bottom: 0.5em;
    background-image: linear-gradient(transparent 65%, 
    rgba(0, 255, 0, 1) 65%, rgba(0, 255, 0, 1) 70%,
    rgba(0, 255, 0, 0) 70%, rgba(0, 255, 0, 0) 75%,
    rgba(255, 0, 0, 1) 75%, rgba(255, 0, 40, 1) 80%,
    transparent 80%, transparent 85%,
    rgba(255, 0, 255, 1) 85%, rgba(255, 0, 255, 1) 90%,
    transparent 90%, transparent 95%,
    rgba(0, 0, 255, 1) 95%, rgba(0, 0, 255, 1) 100%,
    transparent 0%
    );
    
```

And this is what the final CSS on the relevant span would look like:

```

    padding-bottom: 0.5em;
    background-image: linear-gradient(rgba(121, 85, 72, 0.3) 20%, 
    rgba(0, 0, 255, 0.3) 20%, rgba(0, 0, 255, 0.3) 40%, 
    rgba(38, 88, 40, 0.3) 40%, rgba(38, 88, 40, 0.3) 60%,
    transparent 60%, transparent 65%,
    rgba(38, 88, 40, 1) 65%, rgba(38, 88, 40, 1) 70%,
    transparent 70%, transparent 75%,
    rgba(255, 0, 0, 1) 75%, rgba(255, 0, 40, 1) 80%,
    transparent 80%, transparent 85%,
    rgba(255, 0, 0, 1) 85%, rgba(255, 0, 40, 1) 90%,
    transparent 90%, transparent 95%,
    rgba(255, 0, 0, 1) 95%, rgba(255, 0, 40, 1) 100%,
    transparent 0%    );
    ```

**Comment by bvahedy** [comment]
Andrew, I feel we are losing the concept of what an underline's purpose
should be. I feel the purpose of under line is to just bring a lesser
attention than a highlight to a text. Having multiple colored underline
just doesn't make sense. I would like to hear what others think, may be in
too dogmatic about this issue.

**Comment by agrogers** [comment]
Hi Ben. Do mean that if you have added multiple bookmarks with a whole verse underline style you prefer seeing only a single underline?

**Comment by bvahedy** [comment]
No, I mean there should only be one underline allowed on any verse.
Multiple highlight is good, but multiple underline doesn't make any sense.
But that's just me, may be others like multiple underlines, but I don't see
a purpose for it.

**Comment by tuomas2** [comment]
Nice proof of concept. Not sure if "jumping colors between splits" can be easily avoided (due to "black magic" you already faced, needed to implement all this). Otherwise it sounds it is doable. But doesn't bottom padding cause side effects? Does it work properly with different line heights (which is adjustable)?

**Comment by agrogers** [comment]
OK, I think I follow. So for you the solution is simply to add one bookmark
only?

The picture I showed was a bit ugly and overpowering. I would never do it
like that in practice.

I like the underline because I don't like a whole verse having a background
color. I definitely have more then one whole-verse bookmarks on a verses.
But what I would do more often with multiple colored underlines is mark
individual words and phrases with underlines rather than the back color
that I have to use now. I like options 😊

**Comment by agrogers** [comment]
Yeah, I checked different line heights and I was surprised that it did seem
to work well.

**Comment by bvahedy** [comment]
Ok, I understand Andrew, but I fear having too many options may bite us in
the ass in the long run.

**Comment by agrogers** [comment]
@Dleemuse, Tuomas has got the multi-underline working now and it is available in the latest beta.

**Comment by Dleemuse** [comment]
Thank you, just downloaded the newest version last night, going to check it out now! 
D. 

Sent from Yahoo Mail on Android 
 
  On Fri, Aug 13, 2021 at 1:13 AM, Tuomas ***@***.***> wrote:   


Closed #1407.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by Dleemuse** [comment]
Thank you so much for the multi underline feature. This is probably overkill for some peoe, but I love being able at a glance to see how many bookmarks, & which boommark(s) are of extended length. 




Debra

Sent from Yahoo Mail on Android 
 
  On Fri, Aug 13, 2021 at 6:41 AM, Dlee ***@***.***> wrote:   Thank you, just downloaded the newest version last night, going to check it out now! 
D. 

Sent from Yahoo Mail on Android 
 
  On Fri, Aug 13, 2021 at 1:13 AM, Tuomas ***@***.***> wrote:   


Closed #1407.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.
